File: framework/languages/java/e2e/ResilienceLifecycle/Client/src/main/java/systems/zlink/e2e/resiliencelifecycle/client/Scenarios/RlD4MissingRequestHandlerScenario.java
package systems.zlink.e2e.resiliencelifecycle.client.Scenarios;

import systems.zlink.e2e.resiliencelifecycle.client.Support.ResilienceScenarioContext;
import java.time.Duration;
import systems.zlink.e2e.resiliencelifecycle.shared.Contracts;

public final class RlD4MissingRequestHandlerScenario {
    private RlD4MissingRequestHandlerScenario() {
    }

    public static void run(ResilienceScenarioContext context) {
        try {
            context.requestUnhandled("d4-missing-handler", Duration.ofSeconds(3));
            throw new IllegalStateException("RL-D4 missing handler request unexpectedly completed");
        } catch (RuntimeException expected) {
            context.waitForDispatchErrorAny("UnhandledReq", context.adminA(), context.adminB());
        }
        Contracts.WorkRes followUp = context.request("d4-follow-up", Duration.ofSeconds(3));
        ResilienceScenarioContext.ensure("work:d4-follow-up".equals(followUp.value()),
            "RL-D4 follow-up payload mismatch");
        System.out.println("scenario RL-D4 passed");
    }
}

File: framework/languages/java/e2e-kotlin/ResilienceLifecycle/Client/src/main/kotlin/systems/zlink/e2e/kotlin/resiliencelifecycle/Scenarios/RlD4MissingRequestHandlerScenario.kt
package systems.zlink.e2e.kotlin.resiliencelifecycle

fun ClientScenarioContext.runMissingRequestHandlerScenario() {
    val result = requestUnhandledRaw("d4-missing-handler")
    ensure(result.status >= 500, "RL-D4 missing handler request unexpectedly succeeded with ${result.status}")
    ensure(
        result.body.contains("HANDLER_MISSING") && result.body.contains("UnhandledReq"),
        "RL-D4 missing handler response did not expose a public request failure: ${result.body}",
    )
    waitForDispatchErrorAny("UnhandledReq", adminA(), adminB())

    val followUp = requestWork("d4-follow-up")
    ensure(followUp.value() == "work:d4-follow-up", "RL-D4 follow-up payload mismatch")
    println("scenario RL-D4 passed")
}
